OSDOCS-14879: add enablement procedure (#94480)
openshift/openshift-docs: modules/nw-ingress-gateway-api-enable.adoc

:_mod-docs-content-type: PROCEDURE
[id="nw-ingress-gateway-api-enable_{context}"]
= Getting started with Gateway API for the Ingress Operator

When you enable Gateway API, it installs the subscription, the Istio operator, the custom resource definitions (CRDs), and creates the Istio resources.

.Procedure

. Create a `GatewayClass` object that contains the following information:
+
.Example `GatewayClass` CR
[source,yaml]
----
apiVersion: gateway.networking.k8s.io/v1
kind: GatewayClass
metadata:
  name: openshift-default
spec:
  controllerName: openshift.io/gateway-controller/v1 <1>
----
<1> The controller name.
+
[IMPORTANT]
====
The controller name must be exactly as shown for the Ingress Operator to manage it. If you set this field to anything else, the Ingress Operator ignores the `GatewayClass` object and all associated `Gateway`, `GRPCRoute`, and `HTTPRoute` objects. The controller name is tied to the implementation of Gateway API in {product-title}, and `openshift.io/gateway-controller/v1` is the only controller name allowed.
====

. Run the following command to create the `GatewayClass` resource:
+
[source,terminal]
----

> $ oc create -f <gatewayclass_resource>.yaml

Reviewer: <gatewayclass_resource>.yaml isn't defined anywhere.

// Modules included in the following assemblies:
//
// * networking/configuring_ingress_cluster_traffic/ingress-gateway-api.adoc

:_mod-docs-content-type: PROCEDURE
[id="nw-ingress-gateway-api-enable_{context}"]
= Getting started with Gateway API for the Ingress Operator

When you enable Gateway API, it installs the subscription, the Istio operator, the custom resource definitions (CRDs), and creates the Istio resources.

> 

Reviewer: Suggestion in https://github.com/openshift/openshift-docs/pull/94480/files#r2145713754

istiod-openshift-gateway   1/1     1            1           55s
router-default             2/2     2            2           6h4m
----

. Create a secret by using the default certificate by running the following command:
+
[source,terminal]
----
$ oc -n openshift-ingress create secret tls gwapi-wildcard --cert=wildcard.crt --key=wildcard.key
----

. Get the domain of the Ingress Operator by running the following command:
+
[source,terminal]
----
$ DOMAIN=$(oc get ingresses.config/cluster -o jsonpath={.spec.domain})
----

. Create a `Gateway` object:

.. Create a YAML file, `example-gateway.yaml`, that contains the following information:
+
.Example `Gateway` CR
[source,yaml]
----
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: example-gateway
  namespace: openshift-ingress <1>
spec:
  gatewayClassName: openshift-default <2>
  listeners:
  - name: https <3>
    hostname: "*.gwapi.${DOMAIN}" <4>
    port: 443
    protocol: HTTPS
    tls:
      mode: Terminate
      certificateRefs:
      - name: gwapi-wildcard <5>
    allowedRoutes:
      namespaces:
        from: All
----
<1> The `Gateway` object must be created in the `openshift-ingress` namespace.

> <2> The `Gateway` object must reference the name of the previously created `GatewayClass` object.

Reviewer: In the Getting Started guide, it's ok to get them started without extra work for DNS.

<3> The HTTPS listener listens for HTTPS requests that match a subdomain of the cluster domain. You use this listener to configure ingress to your applications by using Gateway API `HTTPRoute` resources.
<4> The hostname must be a subdomain of the Ingress Operator domain. If you use a domain, the listener tries to serve all traffic in that domain.
<5> The name of the previously created secret using the default certificate.

.. Apply the resource by running the following command:
+
[source,terminal]
----
$ oc apply -f example-gateway.yaml
----

.. Optional: When you create a `Gateway` object, {SMProductName} automatically provisions a deployment and service with the same name. Verify this by running the following commands:
+
[source,terminal]
----
$ oc get deployment -n openshift-ingress example-gateway-openshift-default
----
+
[source,terminal]
----
$ oc get service -n openshift-ingress example-gateway-openshift-default
----
+
.Example output
[source,terminal]
----
NAME              READY   UP-TO-DATE   AVAILABLE   AGE
example-gateway   1/1     1            1           25s

> ----

Reviewer 2: <img width="736" alt="image" src="https://github.com/user-attachments/assets/2810fd93-f00b-49e0-89d7-3f1b10c2446f" />

> ----

Reviewer: They should both be named "example-gateway-openshift-default" in the output.

+
.Example output
[source,terminal]
----
NAME              TYPE           CLUSTER-IP   EXTERNAL-IP         PORT(S)      AGE
example-gateway   LoadBalancer   10.1.2.3     <external_ipname>   <port_info>  47s

> ----

Reviewer: @kalexand-rh @jldohmann is there any way this can be fixed before docs freeze?

.. Optional: The Ingress Operator automatically creates a `DNSRecord` CR using the hostname from the listeners, and adds the label `istio.io/gateway-name=example-gateway`. Verify the status of the DNS record by running the following command:
+
[source,terminal]
----

> $ oc get dnsrecord -n openshift-ingress -l istio.io/gateway-name=example-gateway -o yaml

Reviewer: https://github.com/openshift/openshift-docs/pull/94480/files#r2145462579 should be here.

  hostnames: ["example.gwapi.${DOMAIN}"] <3>
  rules: <4>
  - backendRefs:
    - name: example-app <5>
      port: 8443
----
<1> The namespace you are deploying your application.
<2> This field must point to the `Gateway` object you previously configured.
<3> The hostname must match the one specified in the `Gateway` object. In this case, the listeners use a wildcard hostname.
<4> This field specifies the backend references that point to your service.
<5> The name of your application.

.. Apply the resource by running the following command:
+
[source,terminal]
----
$ oc apply -f example-route.yaml
----
+
.Example output
[source,terminal]
----
httproute.gateway.networking.k8s.io/http created
----

.Verification

. Verify that the `Gateway` object is deployed and has the condition `Programmed` by running the following command:
+
[source,terminal]
----
$ oc wait -n openshift-ingress --for=condition=Programmed gateways.gateway.networking.k8s.io example-gateway
----
+
.Example output
[source,terminal]
----
gateway.gateway.networking.k8s.io/example-gateway condition met
----

. Send a request to the configured `HTTPRoute` object hostname:
+
[source,terminal]
----

> $ curl -I -cacert <local cert file> https://example.gwapi.${DOMAIN}:443

Reviewer: Hm, the "-k" means don't validate certs.  Is it really needed?

> $ curl -I -cacert <local cert file> https://example.gwapi.${DOMAIN}:443

Reviewer 3: I noticed that in my cluster without the -k I get error:
```
curl: (60) SSL: certificate subject name '*.apps.iamin-aws0613.qe.devcluster.openshift.com' does not match target host name 'example.gwapi.apps.iamin-aws0613.qe.devcluster.openshift.com'
More details here: https://curl.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
```

> $ curl -I -cacert <local cert file> https://example.gwapi.${DOMAIN}:443

Reviewer 3: -k not needed

> $ curl -I -cacert <local cert file> https://example.gwapi.${DOMAIN}:443

Reviewer: @kalexand-rh the "--cacert" is needed, please accept @rhamini3 's suggestion.

----
$ DOMAIN=$(oc get ingresses.config/cluster -o jsonpath={.spec.domain})
----

. Create a `Gateway` object:

.. Create a YAML file, `example-gateway.yaml`, that contains the following information:
+
.Example `Gateway` CR
[source,yaml]
----
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: example-gateway
  namespace: openshift-ingress <1>
spec:
  gatewayClassName: openshift-default <2>
  listeners:
  - name: https <3>
    hostname: "*.gwapi.${DOMAIN}" <4>
    port: 443
    protocol: HTTPS
    tls:
      mode: Terminate
      certificateRefs:
      - name: gwapi-wildcard <5>
    allowedRoutes:
      namespaces:
        from: All
----
<1> The `Gateway` object must be created in the `openshift-ingress` namespace.
<2> The `Gateway` object must reference the name of the previously created `GatewayClass` object.
<3> The HTTPS listener listens for HTTPS requests that match a subdomain of the cluster domain. You use this listener to configure ingress to your applications by using Gateway API `HTTPRoute` resources.
<4> The hostname must be a subdomain of the Ingress Operator domain. If you use a domain, the listener tries to serve all traffic in that domain.

> <5> The name of the previously created secret using the default certificate.

Reviewer: This can be handled in a follow up.
```suggestion
<5> The name of the previously created secret.
```

.. Apply the resource by running the following command:
+
[source,terminal]
----
$ oc apply -f example-gateway.yaml
----

.. Optional: When you create a `Gateway` object, {SMProductName} automatically provisions a deployment and service with the same name. Verify this by running the following commands:
*** To verify the deployment, run the following command:
+
[source,terminal]
----
$ oc get deployment -n openshift-ingress example-gateway-openshift-default
----
+
.Example output
[source,terminal]
----
NAME              READY   UP-TO-DATE   AVAILABLE   AGE

> example-gateway   1/1     1            1           25s

Reviewer: ```suggestion
example-gateway-openshift-default   1/1     1            1           25s
```

----
$ oc get service -n openshift-ingress example-gateway-openshift-default
----
+
.Example output
[source,terminal]
----
NAME                                TYPE           CLUSTER-IP   EXTERNAL-IP         PORT(S)      AGE
example-gateway-openshift-default   LoadBalancer   10.1.2.3     <external_ipname>   <port_info>  47s
----

.. Optional: The Ingress Operator automatically creates a `DNSRecord` CR using the hostname from the listeners, and adds the label `gateway.networking.k8s.io/gateway-name=example-gateway`. Verify the status of the DNS record by running the following command:
+
[source,terminal]
----
$ oc -n openshift-ingress get dnsrecord -l gateway.networking.k8s.io/gateway-name=example-gateway -o yaml
----
+
.Example output
[source,yaml]
----
kind: DNSRecord
  ...
status:
  ...
  zones:
  - conditions:
    - message: The DNS provider succeeded in ensuring the record
      reason: ProviderSuccess
      status: "True"
      type: Published
    dnsZone:
      tags:
        ...
  - conditions:
    - message: The DNS provider succeeded in ensuring the record
      reason: ProviderSuccess
      status: "True"
      type: Published
    dnsZone:
      id: ...
----

> . Create an `HTTPRoute` resource that directs requests to your application:

Reviewer: ```suggestion
. Create an `HTTPRoute` resource that directs requests to your already-created namespace and application called `example-app/example-app`:
```

.. Create a YAML file, `example-route.yaml`, that contains the following information:
+
.Example `HTTPRoute` CR
[source,yaml]
----
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: example-route
  namespace: example-app-ns <1>
spec:
  parentRefs: <2>
  - name: example-gateway
    namespace: openshift-ingress
  hostnames: ["example.gwapi.${DOMAIN}"] <3>

>   rules: <4>
>   - backendRefs:

Reviewer: ```suggestion
  rules:
  - backendRefs: <4>
```

// Modules included in the following assemblies:
//
// * networking/configuring_ingress_cluster_traffic/ingress-gateway-api.adoc

:_mod-docs-content-type: PROCEDURE
[id="nw-ingress-gateway-api-enable_{context}"]

> = Enabling Gateway API for the Ingress Operator

Reviewer: Is this under a section for Gateway API?  I think a better title might be "Getting Started with Gateway API".  But I guess it depends on the surrounding content.

When you enable Gateway API, it installs the subscription, the Istio operator, the custom resource defintions (CRDs), and creates the Istio resources. The following procedure describes how to enable Gateway API.

.Procedure

> . Create a `GatewayClass` object with the following manifest:

Author: no, you make a good point. i've got an update that rewords things to be more inline with some of the content i see when i search the repo

+
[source,yaml]
----
apiVersion: gateway.networking.k8s.io/v1
kind: GatewayClass
metadata:
  name: openshift-default
spec:
  controllerName: openshift.io/gateway-controller/v1 <1>
----

> <1> The controller name must be `v1` for the Ingress Operator to manage it. Setting this field to anything else will be ignored by the Ingress Operator, because it is assumed to be a different implementation.

Reviewer: ```suggestion
<1> The controller name must be exactly as shown for the Ingress Operator to manage it. Setting this field to anything means the GatewayClass and all of its associated Gateways, GRPCRoutes, and HTTPRoutes will be ignored by the Ingress Operator.  The controllerName is tied to the implementation and `openshift.io/gateway-controller/v1` is the only controllerName that can be used with the OpenShift implementation.
```

Reviewer: Actually, this little note should be highlighted as an "Important" note.  We don't want anyone to miss this.

Author: I think it's important to keep the context, so the callout will stand, but we can add an `IMPORTANT` admonition in a way that is chained to the callout for extra visibility

> <1> The controller name must be `v1` for the Ingress Operator to manage it. Setting this field to anything else will be ignored by the Ingress Operator, because it is assumed to be a different implementation.

Reviewer 3: i mean to say that the status of the gatewayclass will be unknown and not be acepted
```
  status:
    conditions:
    - lastTransitionTime: "1970-01-01T00:00:00Z"
      message: Waiting for controller
      reason: Pending
      status: Unknown  <--
      type: Accepted
```

> <1> The controller name must be `v1` for the Ingress Operator to manage it. Setting this field to anything else will be ignored by the Ingress Operator, because it is assumed to be a different implementation.

Author: gotcha. thank you, i was going to ask how a user would find this out but it looks like Candace mentioned it here: https://github.com/openshift/openshift-docs/pull/94480#discussion_r2140752939

> 

Author: actually this seems like it happens after creating the `GatewayClass` resource, is that right? so maybe it's more like a verification step

> 

Reviewer: Yes, it verifies that the behind-the-scenes activities took place.

. Create a secret using the default certificate:
+
[source,terminal]
----

> $ oc -n openshift-ingress create secret tls gwapi-wildcard --cert=wildcard.crt --key=wildcard.key

Reviewer: Does it tell them anywhere how to create the wildcard.crt and wildcard.key?

Author: we have a lot of existing docs on creating certificates. do you mean something like this? https://docs.redhat.com/en/documentation/openshift_container_platform/4.18/html/security_and_compliance/configuring-certificates i can link it

Reviewer 4: In the [tech preview guides](https://github.com/openshift/network-edge-tools/blob/45a94c03270ceaadf5d204a592645ed5d297de1a/docs/blogs/IntroGatewayAPI/GettingCreative.md?plain=1#L229C1-L230C118) we had this bit of context about the creation of a secret:
```
Certificates are handled differently in Gateway API compared to OpenShift Route objects. Gateway API uses references to
Kubernetes Secret objects whereas the OpenShift Route API embeds the certificate and key strings in the Route object.
```
WDYT about including it here @candita?

Author: @grzpiotrowski @candita so to clarify, is the request to add that snippet within this doc or create a new procedure for handling secrets for Gateway API?

----

. Create a `Gateway` object with the following manifest:
+
[source,yaml]
----
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: example-gateway
  namespace: openshift-ingress
spec:
  gatewayClassName: openshift-default <1>
  listeners: <2>
  - name: http

>     hostname: "*.gwapi.example.com"

Reviewer: Including the correct domain will be super helpful:
```suggestion
    hostname: "*.gwapi.${DOMAIN}"
```

    port: 80
    protocol: HTTP
    allowedRoutes:
      namespaces:

>         from: All

Reviewer: Please remove this section.  A single listener will be a better example to start with.

PS - can you tell me where you got this example from?

  - name: https

>     hostname: "*.gwapi.example.com"

Reviewer: ```suggestion
    hostname: "*.gwapi.${DOMAIN}"
```

    port: 443
    protocol: HTTPS
    tls:
      mode: Terminate
      certificateRefs:

>       - name: gwapi-wildcard <3>

Author: 🤦 lol ofc i'd forget to add it. thank you, i'll update this with the callout

    allowedRoutes:
      namespaces:
        from: All

> ----

Reviewer: ```suggestion
----
EOF
```

<1> The name of the previously created `GatewayClass` object.

> <2> The associated listeners for the `Gateway` object. These listeners are for HTTP and HTTPS, and let you configure ingress to your applications using Gateway API `HTTPRoute` resources.

Reviewer: Add:

By default, when a Gateway object is created, OSSM will automatically provision a gateway deployment and service with the same name.  You can verify this:

```
$ oc get deployment -n openshift-ingress example-gateway
NAME           READY   UP-TO-DATE   AVAILABLE   AGE
example-gateway   1/1     1            1           25s

$ oc get service -n openshift-ingress example-gateway
NAME           TYPE           CLUSTER-IP       EXTERNAL-IP                          PORT(S)                        AGE
example-gateway   LoadBalancer   10.1.2.3   <external ipname>   <port-info>  47s
```
title: Verifying DNSRecord for a Gateway

The Ingress Operator will automatically create a DNSRecord CR with the hostname from the listeners, and a label of `istio.io/gateway-name=example-gateway`.  Verify the status of the DNSRecord:

```
$ oc get dnsrecord -n openshift-ingress -l istio.io/gateway-name=example-gateway -o yaml
kind: DNSRecord
  ...
status:
  ...
  zones:
  - conditions:
    - message: The DNS provider succeeded in ensuring the record
      reason: ProviderSuccess
      status: "True"
      type: Published
    dnsZone:
      tags:
        ...
  - conditions:
    - message: The DNS provider succeeded in ensuring the record
      reason: ProviderSuccess
      status: "True"
      type: Published
    dnsZone:
      id: ...
```

title: Creating an HTTPRoute to connect the parent Gateway to the backend Service

After following the steps to create a Gateway, create and deploy your application in a new namespace and use it in the next steps.  In the next steps we refer to the application as `example-app` in namespace `example-app-ns`.

Create the HTTPRoute CR that will direct requests for example.gwapi.${DOMAIN} to our new example-app application. Note that the `parentRefs` field points to the Gateway, `rules` specifies backendRefs pointing to our service, and `hostnames` specifies a host name that matches the Gateway’s listeners' wildcard host name:

```
$ oc apply -f - <<EOF
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: example-route
  namespace: example-app-ns
spec:
  parentRefs:
  - name: example-gateway
    namespace: openshift-ingress
  hostnames: ["example.gwapi.${DOMAIN}"]
  rules:
  - backendRefs:
    - name: example-app
      port: 8080
    - name: example-app-https
      port: 8443
EOF

httproute.gateway.networking.k8s.io/http created
```

title: Testing the example

After following all the steps to add a Gateway, HTTPRoute, and application, verify that the gateway deployed and has the condition "Programmed":

```
$ oc wait -n openshift-ingress --for=condition=Programmed gateways.gateway.networking.k8s.io example-gateway
gateway.gateway.networking.k8s.io/example-gateway condition met
```

Send a request to the configured HTTPRoute hostname:

```
$ curl -I http://example.gwapi.${DOMAIN}
```

You should see the response from your application example-app.
